File: JavaScript/8-bitwise.js
'use strict';

const a = 9;
const b = 14;
const c = -9;

const aBinary = a.toString(2);
const bBinary = b.toString(2);
const cBinary = c.toString(2);

console.log(`${a} to base 2:  ${aBinary}`);
console.log(`${b} to base 2: ${bBinary}`);

console.log('Bitwise operators');

const andResult = a & b;
console.log(`${a} & ${b} = ${andResult}`);
console.log(`${aBinary} & ${bBinary} = ${andResult.toString(2)}`);

const orResult = a | b;
console.log(`${a} | ${b}`, ' = ', orResult);
console.log(`${aBinary} | ${bBinary} = ${orResult.toString(2)}`);

const xorResult = a ^ b;
console.log(`${a} ^ ${b}`, ' = ', xorResult);
console.log(`${aBinary} ^ ${bBinary} = ${xorResult.toString(2)}`);

const notA = ~a;
console.log(`~${a} = ${notA}`);
console.log(`~${aBinary} = ${notA.toString(2)}`);

const aShiftLeft = a << 2;
console.log(`${a} << ${2} = `, aShiftLeft);
console.log(`${aBinary} << ${2} = ${aShiftLeft.toString(2)}`);

const bShiftRight = b >> 2;
const bShiftUnsigned = b >>> 2;
console.log(`${b} >> ${2} = `, bShiftRight);
console.log(`${b} >>> ${2} = `, bShiftUnsigned);
console.log(`${bBinary} >> ${2} = ${bShiftRight.toString(2)}`);
console.log(`${bBinary} >>> ${2} = ${bShiftUnsigned.toString(2)}`);

const cShiftRight = c >> 2;
const cShiftUnsigned = c >>> 2;
console.log(`${c} >> ${2} = `, cShiftRight);
console.log(`${c} >>> ${2} = `, cShiftUnsigned);
console.log(`${cBinary} >> ${2} = ${cShiftRight.toString(2)}`);
console.log(`${cBinary} >>> ${2} = ${cShiftUnsigned.toString(2)}`);

File: Haskell/8-bitwise.hs
import Data.Char (chr, ord)
import Data.Bits ((.&.), (.|.), xor, complement, shift, shiftL, shiftR, rotate)
import Numeric (showIntAtBase, showSigned)
import Text.Printf(printf)

main = do
  let toBinary :: Int -> String
      toBinary n = showSigned (showIntAtBase 2 toBinDigit) 0 n ""
        where toBinDigit d = chr (ord '0' + d)

  let a = 9
      b = 14
      c = -9

  let aBinary = toBinary a
      bBinary = toBinary b
      cBinary = toBinary c

  printf "%v to base 2: %v\n" a aBinary
  printf "%v to base 2: %v\n" b bBinary

  print "Bitwise operators"

  printf "%v .&. %v = %v\n"   a       b $ a .&. b
  printf "%v .&. %v = %v\n"   aBinary bBinary $ toBinary $ a .&. b

  printf "%v .|. %v = %v\n"   a       b $ a .|. b
  printf "%v .|. %v = %v\n"   aBinary bBinary $ toBinary $ a .|. b

  printf "%v `xor` %v = %v\n" a       b $ a `xor` b
  printf "%v `xor` %v = %v\n" aBinary bBinary $ toBinary $ a `xor` b

  printf "complement %v = %v\n"    a $ complement a
  printf "complement %v = %v\n"    aBinary $ toBinary $ complement a

  printf "%v `shiftL` 2 = %v\n"    a $ a `shiftL` 2
  printf "%v `shiftL` 2 = %v\n"    aBinary $ toBinary $ a `shiftL` 2

  printf "%v `shiftR` 2 = %v\n"    b $ b `shiftR` 2
  printf "%v `shiftR` 2 = %v\n"    bBinary $ toBinary $ b `shiftR` 2

  printf "%v `shift` 2 = %v\n"     a $ a `shift` 2
  printf "%v `shift` 2 = %v\n"     aBinary $ toBinary $ a `shift` 2

  printf "%v `shift` (-2) = %v\n"  b $ b `shift` (-2)
  printf "%v `shift` (-2) = %v\n"  bBinary $ toBinary $ b `shift` (-2)

  printf "%v `rotate` 2 = %v\n"    b $ b `rotate` 2
  printf "%v `rotate` 2 = %v\n"    b $ toBinary $ b `rotate` 2

  printf "%v `rotate` (-2) = %v\n" c $ c `rotate` (-2)
  printf "%v `rotate` (-2) = %v\n" c $ toBinary $ c `rotate` (-2)
